core/src/test/kotlin/generator/DokkaTestGenerator.kt

```diff
@@ -0,0 +1,101 @@
+package org.jetbrains.dokka
+
+import org.jetbrains.dokka.model.Module
+import org.jetbrains.dokka.pages.ModulePageNode
+import org.jetbrains.dokka.pages.PlatformData
+import org.jetbrains.dokka.plugability.DokkaContext
+import org.jetbrains.dokka.plugability.single
+import org.jetbrains.dokka.renderers.FileWriter
+import org.jetbrains.dokka.utilities.DokkaLogger
+
+class DokkaTestGenerator(
+    private val configuration: DokkaConfiguration,
+    private val logger: DokkaLogger,
+    private val checks: Map<Stage, (State) -> Boolean> = mapOf()
+) : DokkaGenerator(configuration, logger) {
+    companion object {
+        enum class Stage {
+            SETUP, PLUGIN_INIT, DOCUMENTABLES_CREATION, DOCUMENTABLES_MERGING, DOCUMENTABLES_TRANSFORMATION, PAGES_CREATION, PAGES_TRANSFORMATION, RENDERING
+        }
+    }
+
+    data class State(var platforms: Map<PlatformData, EnvironmentAndFacade> = mapOf(),
+                     var context: DokkaContext? = null,
+                     var modulesFromPlatforms: List<Module> = listOf(),
+                     var documentationModel: Module? = null,
+                     var transformedDocumentation: Module? = null,
+                     var pages: ModulePageNode? = null,
+                     var transformedPages: ModulePageNode? = null)
```

Comment:
Still nullable vars that can be avoided easily by changing `enum class Stage` to specific classes and passing the result explicitly to each stage

Reply:
Bump

Reply:
Current version uses lambdas passed as parameters to testing function.